ocaml/typing/typecore.ml

```diff
@@ -5153,6 +5170,118 @@ and type_andops env sarg sands expected_ty =
   let let_arg, rev_ands = loop env sarg (List.rev sands) expected_ty in
   let_arg, List.rev rev_ands
 
+  and type_extension ~loc ~env ~ty_expected ~sexp = function
+  | Extensions.Eexp_list_comprehension (sbody, comp_typell) ->
```

Comment:
I would be weird to have the same struct here but not in the Typedtree. I think it makes sense to have two different structs in the the typedtree.

Reply:
There is possibly a discussion about whether we should
have several constructors, or one constructor with a
nested component indicating whether it is an array or
list comprehension (and, tomorrow, big array, string, etc).

However, this is not what my poorly-worded comment
was aiming at. I was wondering whether it would be worth
the trouble to factor out the common parts in the processing
of `Eexp_list_comprehension` and `Eexp_arr_comprehension`.
Something along the lines of:
```
  and type_extension ~loc ~env ~ty_expected ~sexp = function
  | Extensions.Eexp_list_comprehension (sbody, comp_typell) ->
      type_comprehension ~loc ~env ~ty_expected ~sexp ~sbody ~comp_typell
        ~container_type:Predef.type_list ~build:(fun body comp_type ->
            Texp_list_comprehension (body, comp_type))
  | Extensions.Eexp_arr_comprehension (sbody, comp_typell) ->
      type_comprehension ~loc ~env ~ty_expected ~sexp ~sbody ~comp_typell
        ~container_type:Predef.type_array ~build:(fun body comp_type ->
            Texp_arr_comprehension (body, comp_type))

  and type_comprehension ~loc ~env ~ty_expected ~sexp    ~sbody ~comp_typell ~container_type ~build =
    if !Clflags.principal then begin_def ();
    let without_cont_ty = Ctype.newvar ()  in
    unify_exp_types loc env
      (instance (container_type without_cont_ty)) (instance ty_expected);
    if !Clflags.principal then begin
      end_def();
      generalize_structure without_cont_ty;
    end;
    let comp_type, new_env =
      type_comprehension_list ~loc ~env ~container_type ~comp_typell
    in
    let body = type_expect new_env sbody (mk_expected without_cont_ty) in
    re {
      exp_desc = build body comp_type;
      exp_loc = loc; exp_extra = [];
      exp_type = instance (container_type body.exp_type);
      exp_attributes = sexp.pexp_attributes;
      exp_env = env }
```

Reply:
Oh yes! That looks very good to me!
Done.